src/daft-dsl/src/functions/python/mod.rs

```diff
@@ -103,13 +114,121 @@ pub fn stateful_udf(
     name: &str,
     expressions: &[ExprRef],
     return_dtype: DataType,
+    resource_request: Option<ResourceRequest>,
 ) -> DaftResult<Expr> {
     Ok(Expr::Function {
         func: super::FunctionExpr::Python(PythonUDF::Stateful(StatefulPythonUDF {
             name: name.to_string().into(),
             num_expressions: expressions.len(),
             return_dtype,
+            resource_request,
         })),
         inputs: expressions.into(),
     })
 }
+
+/// Replaces resource_requests on UDF expressions in the provided expression tree
+pub fn replace_udf_resource_request(
+    expr: ExprRef,
+    override_resource_request: &ResourceRequest,
+) -> ExprRef {
+    expr.transform(|e| match e.as_ref() {
+        Expr::Function {
+            func:
+                FunctionExpr::Python(PythonUDF::Stateful(
+                    original @ StatefulPythonUDF {
+                        resource_request, ..
+                    },
+                )),
+            inputs,
+        } => {
+            if let Some(existing_rr) = resource_request
+                && existing_rr == override_resource_request
+            {
+                return Ok(Transformed::no(e));
+            }
+            let new_expr = Expr::Function {
+                func: FunctionExpr::Python(PythonUDF::Stateful(StatefulPythonUDF {
+                    resource_request: Some(override_resource_request.clone()),
+                    ..original.clone()
+                })),
+                inputs: inputs.clone(),
+            };
+            Ok(Transformed::yes(new_expr.arced()))
+        }
+        Expr::Function {
+            func:
+                FunctionExpr::Python(PythonUDF::Stateless(
+                    original @ StatelessPythonUDF {
+                        resource_request, ..
+                    },
+                )),
+            inputs,
+        } => {
+            if let Some(existing_rr) = resource_request
+                && existing_rr == override_resource_request
+            {
+                return Ok(Transformed::no(e));
+            }
+            let new_expr = Expr::Function {
+                func: FunctionExpr::Python(PythonUDF::Stateless(StatelessPythonUDF {
+                    resource_request: Some(override_resource_request.clone()),
+                    ..original.clone()
+                })),
+                inputs: inputs.clone(),
+            };
+            Ok(Transformed::yes(new_expr.arced()))
+        }
+        _ => Ok(Transformed::no(e)),
+    })
+    .unwrap()
+    .data
+}
+
+/// Generates a ResourceRequest by inspecting an iterator of expressions.
+/// Looks for ResourceRequests on UDFs in each expression presented, and merges ResourceRequests across all expressions.
+pub fn get_resource_request(exprs: &[ExprRef]) -> Option<ResourceRequest> {
```

Comment:
Nit, maybe for another PR: if I'm not wrong, I don't see tests for resource request merging, but this would be a good one to test.